### PR TITLE
Upgrades changes

### PR DESCRIPTION
**Pull request by ann-sherin**
This contains changes to the upgrades code like: shorten output field names, change output file format from csv to json, add timer stats, fix errors, timepoint multipliers external catalog

### CODE REVIEW

File: disco/extensions/upgrade_simulation/upgrade_simulation.py [hunk 1]
```diff
@@ -119,39 +126,44 @@ def run(
...
-            output_csv_voltage_upgrades_filepath=self.get_voltage_upgrades_csv_file(),
+            output_json_xfmr_upgrades_filepath=self.get_transformer_upgrades_json_file(),
+            output_json_line_upgrades_filepath=self.get_line_upgrades_json_file(),
+            output_json_voltage_upgrades_filepath=self.get_voltage_upgrades_json_file(),
             cost_database_filepath=cost_database_filepath,
             thermal_cost_output_filepath=self.get_thermal_upgrade_costs_file(),
             voltage_cost_output_filepath=self.get_voltage_upgrade_costs_file(),
             total_cost_output_filepath=self.get_total_upgrade_costs_file()
         )
+    TIMER_STATS.log_stats(clear=True)
```
Comment (daniel-thom) [inline disco/extensions/upgrade_simulation/upgrade_simulation.py:169]: Should this be tabbed to the right once?

Comment (ann-sherin) [reply disco/extensions/upgrade_simulation/upgrade_simulation.py]: Sorry, I hadn't pushed my latest commit. Just did

File: disco/extensions/upgrade_simulation/upgrades/common_functions.py [hunk 2]
```diff
@@ -749,4 +1076,103 @@ def compare_dict(old, new):
...
+    value = raw_dict[object_name][property_name][multiplier_name]
+    This value will need to be assigned to the object and run.
+    This hasnt been used yet.
+    
+    Returns
+    -------
+    dict
+    """
+    field = "load_multipliers"
+    if field == "load_multipliers":
```
Comment (daniel-thom) [inline disco/extensions/upgrade_simulation/upgrades/common_functions.py:1098]: Redundant

File: disco/extensions/upgrade_simulation/upgrades/common_functions.py [hunk 3]
```diff
@@ -749,4 +1076,103 @@ def compare_dict(old, new):
...
+    
+    Returns
+    -------
+    dict
+    """
+    field = "load_multipliers"
+    name_list = list(reformatted_dict.keys())
+    if property_list is None:
+        property_list = list(reformatted_dict[name_list[0]].keys())
+    if field == "load_multipliers":
```
Comment (daniel-thom) [inline disco/extensions/upgrade_simulation/upgrades/common_functions.py:1147]: Redundant with line 1143

File: disco/extensions/upgrade_simulation/upgrades/common_functions.py [hunk 4]
```diff
@@ -749,4 +1076,103 @@ def compare_dict(old, new):
...
+        for name in name_list:
+            dss.Loads.Name(name)
```
Comment (daniel-thom) [inline disco/extensions/upgrade_simulation/upgrades/common_functions.py:1149]: This is a very slow way of iterating. It is better to use the native OpenDSS iteration. Could you flip it easily?

File: disco/extensions/upgrade_simulation/upgrades/cost_computation.py [hunk 5]
```diff
@@ -72,10 +75,11 @@ def compute_all_costs(
 
     total_cost_df = get_total_costs(thermal_cost_df, voltage_cost_df)
 
-    # save files
-    thermal_cost_df.to_csv(thermal_cost_output_filepath, index=False)
-    voltage_cost_df.to_csv(voltage_cost_output_filepath, index=False)
-    total_cost_df.to_csv(total_cost_output_filepath, index=False)
+    # breakpoint()
```
Comment (daniel-thom) [inline disco/extensions/upgrade_simulation/upgrades/cost_computation.py:78]: stale

File: disco/extensions/upgrade_simulation/upgrades/thermal_upgrade_functions.py [hunk 6]
```diff
@@ -328,6 +357,10 @@ def identify_parallel_xfmrs(options=None, row=None, parallel_xfmrs_limit=None):
     num_parallel_xfmrs = int(chosen_option["num_parallel"])
     chosen_option["conns"] = convert_list_string_to_list(chosen_option["conns"])
     chosen_option["kVs"] = convert_list_string_to_list(chosen_option["kVs"])
+    if isinstance(chosen_option["%Rs"], str):
+                    chosen_option["%Rs"] = convert_list_string_to_list(chosen_option["%Rs"])
```
Comment (daniel-thom) [inline disco/extensions/upgrade_simulation/upgrades/thermal_upgrade_functions.py:361]: Is the function misnamed? `convert_string_to_list`?

File: disco/extensions/upgrade_simulation/upgrades/automated_thermal_upgrades.py [hunk 7]
```diff
@@ -17,264 +24,224 @@ def determine_thermal_upgrades(
...
+        if len(overloaded_line_list) > 0:            
             line_commands_list, temp_line_upgrades_df = correct_line_violations(
                 line_loading_df=line_loading_df,
                 line_design_pu=thermal_config["line_design_pu"],
                 line_upgrade_options=line_upgrade_options,
                 parallel_lines_limit=thermal_config["parallel_lines_limit"],
-            )
-            logger.info("Corrected line violations.")
+                external_upgrades_technical_catalog=external_upgrades_technical_catalog,)
+            logger.info(f"Iteration_{iteration_counter}: Corrected line violations.")
```
[the remark this answers is not in the document]
Comment (ann-sherin) [reply disco/extensions/upgrade_simulation/upgrades/automated_thermal_upgrades.py:145]: Oh I see, okay. So with logging always use this format : logger.info("Some text %s", var)  
Got it.

File: disco/extensions/upgrade_simulation/upgrade_simulation.py [hunk 8]
```diff
@@ -10,6 +11,7 @@
 class UpgradeSimulation:
     
     def __init__(self, job, job_global_config, output=OUTPUT_DIR):
+        timer_stats_collector.clear()
```
[the remark this answers is not in the document]
Comment (ann-sherin) [reply disco/extensions/upgrade_simulation/upgrade_simulation.py:14]: Okay, so I'll just take the clear statement off from here then.

File: disco/extensions/upgrade_simulation/upgrades/common_functions.py [hunk 9]
```diff
@@ -399,12 +650,79 @@ def get_all_line_info(compute_loading=True, upper_limit=1.5, ignore_switch=True)
...
+                    loading_df = get_all_line_info_instance(compute_loading=compute_loading, upper_limit=upper_limit, ignore_switch=ignore_switch)
+                elif equipment_type.lower() == "transformer":
+                    deciding_column_name = "max_per_unit_loading"
+                    loading_df = get_all_transformer_info_instance(compute_loading=compute_loading, upper_limit=upper_limit)
+                loading_df.set_index("name", inplace=True)
+                comparison_dict[pv_field+"_"+str(multiplier_name)] = loading_df
+        # compare all dataframe, and create one that contains all worst loading conditions (across all multiplier conditions)
+        loading_df = compare_multiple_dataframes(comparison_dict, deciding_column_name, comparison_type="max")
+    else:
+        raise Exception(f"Undefined multiplier_type {multiplier_type} passed.")
```
[the remark this answers is not in the document]
Comment (ann-sherin) [reply disco/extensions/upgrade_simulation/upgrades/common_functions.py:721]: Okay, I've made a note of this

File: disco/extensions/upgrade_simulation/upgrades/thermal_upgrade_functions.py (hunk 6 above)
Comment (ann-sherin) [reply disco/extensions/upgrade_simulation/upgrades/thermal_upgrade_functions.py:361]: I couldnt think of a better name-- this basically converts "[2,3,4]" to [2,3,4].
Hence the name "list_string" to "list" -- Looks like thats confusing. I guess we could just call it this then : convert_string_to_list

Comment (daniel-thom) [reply disco/extensions/upgrade_simulation/upgrades/thermal_upgrade_functions.py:361]: Ah. Consider `ast.literal_eval()`

File: disco/extensions/upgrade_simulation/upgrades/common_functions.py [hunk 10]
```diff
@@ -327,7 +552,48 @@ def get_all_transformer_info(compute_loading=True, upper_limit=1.5):
...
+    info_dict = {}
+    info_dict["line_placement"] = None
+    if line_series["line_definition_type"] == "geometry":
+            dss.Circuit.SetActiveClass("linegeometry")
+            dss.ActiveClass.Name(line_series["geometry"])
+            h = float(dss.Properties.Value("h"))
+            info_dict["h"] = 0
+            if h >= 0:
+                info_dict["line_placement"] = "overhead"
+            elif h < 0:
```
Comment (jgu2) [inline disco/extensions/upgrade_simulation/upgrades/common_functions.py:584]: `else:`

File: disco/extensions/upgrade_simulation/upgrades/common_functions.py [hunk 11]
```diff
@@ -399,12 +650,79 @@ def get_all_line_info(compute_loading=True, upper_limit=1.5, ignore_switch=True)
...
+            -------
+            Dataframe
+            """
+            summary_df = pd.DataFrame()
+            for df_name in comparison_dict.keys():
+                summary_df[df_name] = comparison_dict[df_name][deciding_column_name]
+            if comparison_type == "max":
+                label_df = summary_df.idxmax(axis=1)  # find dataframe name that has max 
+            elif comparison_type == "min":
+                label_df = summary_df.idxmax(axis=1)  # find dataframe name that has max
```
Comment (jgu2) [inline disco/extensions/upgrade_simulation/upgrades/common_functions.py:673]: typo? `# find dataframe name that has min`

File: disco/extensions/upgrade_simulation/upgrades/common_functions.py [hunk 12]
```diff
@@ -399,12 +650,79 @@ def get_all_line_info(compute_loading=True, upper_limit=1.5, ignore_switch=True)
...
+            loading_df = get_all_line_info_instance(compute_loading=compute_loading, upper_limit=upper_limit, ignore_switch=ignore_switch)
+        elif equipment_type == "transformer":
+            loading_df = get_all_transformer_info_instance(compute_loading=compute_loading, upper_limit=upper_limit)
+        return loading_df
+    if multiplier_type == 'uniform':
+        comparison_dict = {}
+        for pv_field in timepoint_multipliers["load_multipliers"].keys():
+            logger.debug(pv_field)
+            for multiplier_name in timepoint_multipliers["load_multipliers"][pv_field]:
+                logger.debug(multiplier_name)
```
Comment (jgu2) [inline disco/extensions/upgrade_simulation/upgrades/common_functions.py:707]: Remove, or add more words around the debug log?

File: disco/extensions/upgrade_simulation/upgrades/common_functions.py [hunk 13]
```diff
@@ -749,4 +1076,103 @@ def compare_dict(old, new):
...
+            multiplier_list = multiplier_list + value
+        df = dss.utils.class_to_dataframe(object_name)
+        df.reset_index(inplace=True)
+        df['name'] = df['index'].str.split(".", expand=True)[1]
+        name_list = list(df['name'].values)
+        del df["index"]
+        df.set_index('name', inplace=True)
+        raw_dict = {}
+        for property in property_list:
+            logger.debug(property)
```
Comment (jgu2) [inline disco/extensions/upgrade_simulation/upgrades/common_functions.py:1113]: same here.

File: disco/extensions/upgrade_simulation/upgrades/common_functions.py [hunk 14]
```diff
@@ -749,4 +1076,103 @@ def compare_dict(old, new):
...
+            df[property] = df[property].astype(float)
+            new_df = pd.DataFrame(index=name_list, columns=multiplier_list)
+            new_df.index.name = 'name'
+            for multiplier in multiplier_list:
+                logger.debug(multiplier)
```
Comment (jgu2) [inline disco/extensions/upgrade_simulation/upgrades/common_functions.py:1118]: same here.

File: disco/extensions/upgrade_simulation/upgrades/cost_computation.py [hunk 15]
```diff
@@ -244,7 +242,7 @@ def compute_line_costs(line_upgrades_df=None, line_cost_database=None, **kwargs)
         "m": 1,
     }
     # choose which properties are to be saved
-    properties_list = list(set(output_columns_list) - {"type"})
+    # output_properties_list = list(set(output_columns_list) - {"type"})
```
Comment (jgu2) [inline disco/extensions/upgrade_simulation/upgrades/cost_computation.py:245]: Is it commented out?

Comment (ann-sherin) [reply disco/extensions/upgrade_simulation/upgrades/cost_computation.py:245]: Yes, deleted from file

File: disco/extensions/upgrade_simulation/upgrades/common_functions.py (hunk 11 above)
Comment (ann-sherin) [reply disco/extensions/upgrade_simulation/upgrades/common_functions.py:673]: Oh yes, changed.

File: disco/extensions/upgrade_simulation/upgrades/common_functions.py (hunk 10 above)
Comment (ann-sherin) [reply disco/extensions/upgrade_simulation/upgrades/common_functions.py:584]: Yeah, there's no other condition. So changed elif to else.

File: disco/extensions/upgrade_simulation/upgrades/common_functions.py (hunk 2 above)
Comment (ann-sherin) [reply disco/extensions/upgrade_simulation/upgrades/common_functions.py:1098]: True, I added a few more lines, to loop over dict keys and raise exception if key other than this is present.
That was my main intent.

File: disco/extensions/upgrade_simulation/upgrades/common_functions.py (hunk 3 above)
Comment (ann-sherin) [reply disco/extensions/upgrade_simulation/upgrades/common_functions.py:1147]: True, I added a few more lines to raise exception if key other than this is present.
That was my main intent.

File: disco/extensions/upgrade_simulation/upgrades/common_functions.py [hunk 16]
```diff
@@ -233,9 +235,84 @@ def get_dictionary_of_duplicates(df, subset=None, index_field=None):
     return mapping_dict
 
 
-# to get metadata: source bus, feeder hear info, substation xfmr information
+def write_to_json(json_file, filename):
```
[the remark this answers is not in the document]
Comment (ann-sherin) [reply disco/extensions/upgrade_simulation/upgrades/common_functions.py:209]: Done

File: disco/extensions/upgrade_simulation/upgrades/common_functions.py [hunk 17]
```diff
@@ -233,9 +235,84 @@ def get_dictionary_of_duplicates(df, subset=None, index_field=None):
...
+    pydss_volt_var_model 
+
+    Returns
+    -------
+    str
+    """
+    if enable_pydss_solve:
+        scenario = pydss_volt_var_model.control1
+    else:
+        scenario = "no_control"
```
[the remark this answers is not in the document]
Comment (ann-sherin) [reply disco/extensions/upgrade_simulation/upgrades/common_functions.py:236]: Changed to control_mode and pf1 for now. In the future, we could read in name -- I've added in a todo

File: disco/extensions/upgrade_simulation/upgrades/common_functions.py [hunk 18]
```diff
@@ -678,6 +1002,9 @@ def determine_available_xfmr_upgrades(xfmr_loading_df):
                      'XfmrCode', 'XRConst', 'X12', 'X13', 'X23', 'LeadLag',
                      'Core', 'RdcOhms', 'normamps', 'emergamps', 'faultrate', 'pctperm',
                      'basefreq', 'amp_limit_per_phase']
+    # TODO
```
[the remark this answers is not in the document]
Comment (ann-sherin) [reply disco/extensions/upgrade_simulation/upgrades/common_functions.py:976]: added text

File: disco/extensions/upgrade_simulation/upgrades/thermal_upgrade_functions.py (hunk 6 above)
Comment (ann-sherin) [reply disco/extensions/upgrade_simulation/upgrades/thermal_upgrade_functions.py:361]: Oh I just checked -- ast.literal_eval is what I used inside that function :D 
I guess I made that function initially with something else in. Replaced with ast.literal_eval in all usages

File: disco/extensions/upgrade_simulation/upgrades/automated_thermal_upgrades.py [hunk 19]
```diff
@@ -17,264 +24,224 @@ def determine_thermal_upgrades(
...
-                f"during upgrade process"
-            )
-        if len(overloaded_line_list) > 0:
+        iteration_counter < max_upgrade_iteration):
+        line_loading_df = get_thermal_equipment_info(compute_loading=True, upper_limit=thermal_config["line_upper_limit"], 
+                                                    equipment_type="line", ignore_switch=ignore_switch, multiplier_type=multiplier_type,
+                                                    timepoint_multipliers=timepoint_multipliers, **simulation_params)
+        overloaded_line_list = list(line_loading_df.loc[line_loading_df["status"] == "overloaded"]["name"].unique())
+        logger.info(f"Iteration_{iteration_counter}: Determined line loadings.")
+        logger.info(f"Iteration_{iteration_counter}: Number of line violations: {len(overloaded_line_list)}")
```
[the remark this answers is not in the document]
Comment (ann-sherin) [reply disco/extensions/upgrade_simulation/upgrades/automated_thermal_upgrades.py:135]: Is this the comment on the f strings for logger? 
I've made a note of this, and can change it in future

**Review by jgu2 — APPROVED** [review]

**Pull request merged**